pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,350 @@
+			// calico-upgrade service is done. The new calico-upgrade
+			// service will clean the old service up.
+			date := time.Now().Format("2006-01-02")
+			log.Info(fmt.Sprintf("Upgrade is in progress. Upgrade log is in c:\\calico-upgrade.%v.log", date))
+			time.Sleep(3 * time.Second)
+			return
+		}
+	}
+}
+
+// Return the base directory for Calico upgrade service.
+func baseDir() string {
+	dir := filepath.Dir(os.Args[0])
+	return "c:\\" + filepath.Base(dir)
+}
+
+// Return if the monitor service is running as part of Enterprise installation.
+func runningEnterprise() bool {
+	return baseDir() == EnterpriseBaseDir
+}
+
+// Return kubeconfig file path for Calico
+func kubeConfigFile() string {
+	return baseDir() + "\\" + CalicoKubeConfigFile
+}
+
+func uninstall() error {
+	path := filepath.Join(baseDir(), "uninstall-calico.ps1")
+	log.Infof("Start uninstall script %s\n", path)
+	stdout, stderr, err := powershell(path)
+	fmt.Println(stdout, stderr)
+	if err != nil {
+		return err
+	}
+	// After the uninstall completes, move the existing calico-node.exe to
+	// a temporary file. The calico-upgrade service is still running so not
+	// doing this means we cannot replace calico-node.exe with the upgrade.
+	stdout, stderr, err = powershell(fmt.Sprintf(`mv %v\calico-node.exe %v\calico-node.exe.to-be-replaced`, baseDir(), baseDir()))
+	fmt.Println(stdout, stderr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func execScript(script string) error {
+	log.Infof("Start script %s\n", script)
+
+	// This has to be done in a separate process because when the new calico services are started, the existing
+	// calico-upgrade service is removed so the new calico-upgrade service can be started.
+	// However, removing the existing calico-upgrade service means the powershell
+	// process running the upgrade script is killed and the installation is left
+	// incomplete.
+	cmd := fmt.Sprintf(`Start-Process powershell -argument %q -WindowStyle hidden`, script)
+	stdout, stderr, err := powershell(cmd)
+
+	if err != nil {
+		return err
+	}
+	fmt.Println(stdout, stderr)
+	return nil
+}
+
+func verifyImagesSharePathPrefix(first, second string) error {
+	n1, err := image.ParseNamed(first)
+	if err != nil {
+		return err
+	}
+	n2, err := image.ParseNamed(second)
+	if err != nil {
+		return err
+	}
+	// Compare the domain parts (e.g. docker.io) of the image references.
+	if image.Domain(n1) != image.Domain(n2) {
+		return fmt.Errorf("images %q and %q do not share the same domain", first, second)
+	}
+
+	// Split the image path. E.g. if the image is
+	// "docker.io/calico/node:v3.21.0" then the path is "calico/node".
+	n1PathParts := strings.Split(image.Path(n1), "/")
+	n2PathParts := strings.Split(image.Path(n2), "/")
+
+	// Remove the last segment of the image path since it will contain the
+	// component name.
+	n1PathPrefix := n1PathParts[:len(n1PathParts)-1]
+	n2PathPrefix := n2PathParts[:len(n2PathParts)-1]
```

Comment:
I think it is possible the image name is `domain/calico-node:xxx` where n1PathParts has zero length.

Reply:
Thanks, you're right. I fixed this to skip checking the path if the image reference is a short one (in this case the path must be the component name)

Reply:
The compare of short path is fine, but it is always good practice to add check for zero length on both image paths, that will make sure [:len(n1PathParts)-1] not panic. Also we don't need to proceed if the length is different? If n1PathParts has more elements than n2PathParts, it will panic to compare them one by one later. 
```
if (len(n1PathParts) == 0) || (len(n2PathParts) == 0) || (len(n1PathParts) != len(n2PathParts) ) {
  failed
}